File: 25_1/s_18/exo2/sorts/merge.go
package sorts

func MergeSort(items []int) []int {
	n := len(items)
	if n < 2 {
		return append([]int(nil), items...)
	}
	mid := n / 2
	left := MergeSort(items[:mid])
	right := MergeSort(items[mid:])
	return merge(left, right)
}

func merge(a, b []int) []int {
	out := make([]int, 0, len(a)+len(b))
	i, j := 0, 0
	for i < len(a) && j < len(b) {
		if a[i] <= b[j] {
			out = append(out, a[i])
			i++
		} else {
			out = append(out, b[j])
			j++
		}
	}
	out = append(out, a[i:]...)
	out = append(out, b[j:]...)
	return out
}

File: 25_2/s_11/sorting/sorting/merge.py
# Сортировка слиянием (Merge Sort)
# Временная сложность: O(n log n)
# Пространственная сложность: O(n)

def merge(left, right, counters):
    merged = []
    i = j = 0

    while i < len(left) and j < len(right):
        counters["comparisons"] += 1
        if left[i] < right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
            counters["swaps"] += 1

    merged.extend(left[i:])
    merged.extend(right[j:])
    return merged


def merge_sort(arr, counters=None):
    if counters is None:
        counters = {"comparisons": 0, "swaps": 0}
    if len(arr) <= 1:
        return arr, counters["comparisons"], counters["swaps"]

    mid = len(arr) // 2
    left, _, _ = merge_sort(arr[:mid], counters)
    right, _, _ = merge_sort(arr[mid:], counters)
    merged = merge(left, right, counters)

    return merged, counters["comparisons"], counters["swaps"]
